src/BaseMailer.php

```diff
@@ -0,0 +1,284 @@
+        $this->composer = $composer;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @return Composer message composer instance.
+     */
+    public function getComposer(): Composer
+    {
+        return $this->composer;
+    }
+
+    /**
+     * Creates a new message instance and optionally composes its body content via view rendering.
+     *
+     * @param string|array|null $view the view to be used for rendering the message body. This can be:
+     *
+     * - a string, which represents the view name for rendering the HTML body of the email.
+     *   In this case, the text body will be generated by applying `strip_tags()` to the HTML body.
+     * - an array with 'html' and/or 'text' elements. The 'html' element refers to the view name
+     *   for rendering the HTML body, while 'text' element is for rendering the text body. For example,
+     *   `['html' => 'contact-html', 'text' => 'contact-text']`.
+     * - null, meaning the message instance will be returned without body content.
+     *
+     * The view to be rendered can be specified in one of the following formats:
+     *
+     * - a relative view name (e.g. "contact") located under [[viewPath]].
+     *
+     * @param array $params the parameters (name-value pairs) that will be extracted and made available in the view file.
+     * @return MessageInterface message instance.
+     */
+    public function compose($view = null, array $params = []): MessageInterface
+    {
+        $message = $this->createMessage();
+        if ($view === null) {
+            return $message;
+        }
+
+        $this->composer->compose($message, $view, $params);
+
+        return $message;
+    }
+
+    /**
+     * Creates a new message instance.
+     * @return MessageInterface message instance.
+     */
+    protected function createMessage(): MessageInterface
+    {
+        return $this->messageFactory->create($this);
+    }
+
+    /**
+     * Sends the given email message.
+     * This method will log a message about the email being sent.
+     * If [[useFileTransport]] is true, it will save the email as a file under [[fileTransportPath]].
+     * Otherwise, it will call [[sendMessage()]] to send the email to its recipient(s).
+     * Child classes should implement [[sendMessage()]] with the actual email sending logic.
+     * @param MessageInterface $message email message instance to be sent
+     * @return bool whether the message has been sent successfully
+     */
+    public function send(MessageInterface $message): bool
+    {
+        if (!$this->beforeSend($message)) {
+            return false;
+        }
+
+        $address = $message->getTo();
+        if (is_array($address)) {
+            $address = implode(', ', array_keys($address));
+        }
+        $this->logger->info('Sending email "' . $message->getSubject() . '" to "' . $address . '"');
+
+        if ($this->useFileTransport) {
```

Comment:
Ideally the "file transport" concept should be implemented as a separate mailer (available by default right from this package). There will be no need for `if()`s then.

Reply:
i.e. `saveMessage` could be moved to a `FileMailer::sendMessage()`.

Reply:
LGTM, but keep file transport inside `BaseMailer` is convenient to enable/disable by difference environments, only a small number of configuration changes are required:

```php
MailerInterface::class => [
    '__class' => SwiftMailer::class,
    'useFileTransport()' => YII_DEBUG ? true : false, // or merging local configuration.
    // ...
]
```

Reply:
Should be even better with separate mailers:

```php
MailerInterface::class => [
    '__class' => YII_DEBUG ? FileMailer::class : SwiftMailer::class,
]
```